File: store/src/test/java/com/nytimes/android/external/store3/StreamOneKeyTest.java
package com.nytimes.android.external.store3;

import com.nytimes.android.external.store3.base.Fetcher;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.base.impl.StoreBuilder;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import io.reactivex.Maybe;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class StreamOneKeyTest {

    private static final String TEST_ITEM = "test";
    private static final String TEST_ITEM2 = "test2";

    @Mock
    Fetcher<String, BarCode> fetcher;
    @Mock
    Persister<String, BarCode> persister;

    private final BarCode barCode = new BarCode("key", "value");
    private final BarCode barCode2 = new BarCode("key2", "value2");

    private Store<String, BarCode> store;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        store = StoreBuilder.<String>barcode()
                .persister(persister)
                .fetcher(fetcher)
                .open();

        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(TEST_ITEM))
                .thenReturn(Single.just(TEST_ITEM2));

        when(persister.read(barCode))
                .thenReturn(Maybe.<String>empty())
                .thenReturn(Maybe.just(TEST_ITEM))
                .thenReturn(Maybe.just(TEST_ITEM2));

        when(persister.write(barCode, TEST_ITEM))
                .thenReturn(Single.just(true));
        when(persister.write(barCode, TEST_ITEM2))
                .thenReturn(Single.just(true));
    }

    @Test
    public void testStream() {
        TestObserver<String> streamObservable = store.stream(barCode).test();
        //first time we subscribe to stream it will fail getting from memory & disk and instead
        //fresh from network, write to disk and notifiy subscribers
        streamObservable.assertValueCount(1);

        store.clear();
        //fresh should notify subscribers again
        store.fresh(barCode).test().awaitCount(1);
        streamObservable.assertValues(TEST_ITEM, TEST_ITEM2);

        //get for another barcode should not trigger a stream for barcode1
        when(fetcher.fetch(barCode2))
                .thenReturn(Single.just(TEST_ITEM));
        when(persister.read(barCode2))
                .thenReturn(Maybe.empty())
                .thenReturn(Maybe.just(TEST_ITEM));
        when(persister.write(barCode2, TEST_ITEM))
                .thenReturn(Single.just(true));
        store.get(barCode2).test().awaitCount(1);
        streamObservable.assertValueCount(2);
    }
}

File: store/src/test/java/com/nytimes/android/external/store3/StreamOneKeyTest.kt
package com.nytimes.android.external.store3

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.Store
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test

class StreamOneKeyTest {

    val fetcher: Fetcher<String, BarCode> = mock()
    val persister: Persister<String, BarCode> = mock()

    private val barCode = BarCode("key", "value")
    private val barCode2 = BarCode("key2", "value2")

    private val store: Store<String, BarCode> = StoreBuilder.barcode<String>()
            .persister(persister)
            .fetcher(fetcher)
            .open()

    @Before
    fun setUp() = runBlocking<Unit> {
        whenever(fetcher.fetch(barCode))
                .thenReturn(TEST_ITEM)
                .thenReturn(TEST_ITEM2)

        whenever(persister.read(barCode))
                .thenReturn(TEST_ITEM)
                .thenReturn(TEST_ITEM2)

        whenever(persister.write(barCode, TEST_ITEM))
                .thenReturn(true)
        whenever(persister.write(barCode, TEST_ITEM2))
                .thenReturn(true)
    }

    @Test
    fun testStream() = runBlocking<Unit> {
        val streamSubscription = store.stream(barCode).openChannelSubscription()
        try {//stream doesn't invoke get anymore so when we call it the channel is empty
            assertThat(streamSubscription.isEmpty).isTrue()

            store.clear()
            //fresh should notify subscribers again
            store.fresh(barCode)
            assertThat(streamSubscription.poll()).isEqualTo(TEST_ITEM)
//        assertThat(streamObservable.poll()).isEqualTo(TEST_ITEM2)

            //get for another barcode should not trigger a stream for barcode1
            whenever(fetcher.fetch(barCode2))
                    .thenReturn(TEST_ITEM)
            whenever(persister.read(barCode2))
                    .thenReturn(TEST_ITEM)
            whenever(persister.write(barCode2, TEST_ITEM))
                    .thenReturn(true)
            store.get(barCode2)
            assertThat(streamSubscription.isEmpty).isTrue()
        } finally {
            streamSubscription.cancel()
        }
    }

    companion object {
        private const val TEST_ITEM = "test"
        private const val TEST_ITEM2 = "test2"
    }
}
